singularity-eos/closure/mixed_cell_models.hpp

```diff
@@ -647,6 +651,491 @@ class PTESolverRhoT : public mix_impl::PTESolverBase<EOSIndexer, RealIndexer> {
+  return neq * neq                 // jacobian
+         + 4 * neq                 // dx, residual, and sol_scratch
+         + 2 * nmat                // rhobar and u in base
+         + 2 * nmat                // nmat sized arrays in fixed T solver
+         + MAX_NUM_LAMBDAS * nmat; // the cache
+}
+inline size_t PTESolverFixedTRequiredScratchInBytes(const int nmat) {
+  return PTESolverFixedTRequiredScratch(nmat) * sizeof(Real);
+}
+
+template <typename EOSIndexer, typename RealIndexer, typename LambdaIndexer>
+class PTESolverFixedT : public mix_impl::PTESolverBase<EOSIndexer, RealIndexer> {
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::AssignIncrement;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::nmat;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::neq;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::ResidualNorm;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::vfrac_total;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::eos;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::rho;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::vfrac;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::sie;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::temp;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::press;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::rho_total;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::u_total;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::MatIndex;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::jacobian;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::residual;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::dx;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::u;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::rhobar;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::Cache;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::Tnorm;
+
+ public:
+  // template the ctor to get type deduction/universal references prior to c++17
+  template <typename EOS_t, typename Real_t, typename Lambda_t>
+  PORTABLE_INLINE_FUNCTION
+  PTESolverFixedT(const int nmat, EOS_t &&eos, const Real vfrac_tot, const Real T_true,
+                Real_t &&rho, Real_t &&vfrac, Real_t &&sie, Real_t &&temp, Real_t &&press,
+                Lambda_t &&lambda, Real *scratch)
+      : mix_impl::PTESolverBase<EOSIndexer, RealIndexer>(nmat, nmat, eos, vfrac_tot,
+                                                         1.0, rho, vfrac, sie, temp,
+                                                         press, scratch) {
+    dpdv = AssignIncrement(scratch, nmat);
+    vtemp = AssignIncrement(scratch, nmat);
+    Tequil = T_true;
+    Ttemp = T_true;
+    Tnorm = 1.0;
+    // TODO(JCD): use whatever lambdas are passed in
+    /*for (int m = 0; m < nmat; m++) {
+      if (lambda[m] != nullptr) Cache[m] = lambda[m];
+    }*/
```

Comment:
We should probably make an issue about this.

Reply:
I am now calculating the offset from the scratch memory and using it in the calling code. We can still create an issue if you think it necessary.

Reply:
The PTE solver uses a portion of the solver scratch memory to store an internal cache. The address of the first location of the cache, as well as the cache size is known and can be calculated anywhere. I perform this calculation in the scope outside the PTE solve and thus have access to the cache used internally by the PTE solver. Therefore I do not need to allocate additional memory for the cache nor do I need to pass it in to the PTE solve, and even if I did, it would not be used as can be seen in the implementation. If you need more context it may be easier to communicate over the phone.

Reply:
Nah I get it now. Allocating the lambdas in scratch memory is nice, but we should still make an issue because in the future, there could be lambda parameters that are not cache parameters but some actual physical values. (For example a mix fraction for a hybrid or multiphase eos.)

Reply:
I agree, but I think it should be part of a different MR. If I made an issue would that suffice for now?

Reply:
An issue suffices for now. We don't currently have a use-case to test against---let's not waste developer effort. I just want to make sure we keep this in mind going forward.

Reply:
addressed in issue #173